### PR TITLE
Add rustdoc settings menu

### PR DESCRIPTION
**Pull request by GuillaumeGomez**
Fixes #18167.

r? @QuietMisdreavus

### CODE REVIEW

**Comment by ollie27** [comment]
If `settings.html` is supposed to be a shared file then its contents should not depend on the crate that generated it so the output is deterministic when documenting multiple crates at the same time. Currently `settings.html` contains lines like:

```html
<a href='./std_unicode/index.html'><img src='https://www.rust-lang.org/logos/rust-logo-128x128-blk-v2.png' alt='logo' width='100'></a>
```

Perhaps it would be better for the settings to be in a drop-down menu like the themes instead of a separate file.

**Comment by QuietMisdreavus** [comment]
Alternately, if it's going to be in a separate page, is it possible to put it into `librustdoc/html/static/` somehow? Or perhaps using a new template so that the sidebar doesn't link to anything crate-specific?

On the other hand, when i ran it myself, the link gets changed to `javascript:void` at page load time. The sidebar also gets filled with the crate list, so at least there's a means to get back to regular documentation.

**Comment by ollie27** [comment]
Another option would be to use a pop-up like the help pop-up.

**Comment by GuillaumeGomez** [comment]
@ollie27: Check out the changes, I remove the url on the logo. :)

**Comment by GuillaumeGomez** [comment]
And I'd prefer to avoid adding even more JS than necessary. Settings is in its own page, with its own JS and everything is fine as is.

**Comment by ollie27** [comment]
The issue with removing the url with JS is that it doesn't help with bit-for-bit deterministic builds (#34902). Also the logo itself is crate specific, as is the favicon. How about not sharing settings.html and generating it once for each crate next to all.html?

**Comment by GuillaumeGomez** [comment]
Because that's repeated data. I don't really see the issue in here, if the favicon is switched because the user changed it in one crate, then so be it?

**Comment by QuietMisdreavus** [comment]
How about this: We can clone the `Layout` when generating the settings page, to remove the `krate`, and also tweak the layout to render the logo, but *not* the link, when there's a logo but no `krate` to link to. That way, the logo is present, the build is reproducible, and the settings page is still one file for the whole bundle.

**Comment by GuillaumeGomez** [comment]
@QuietMisdreavus: Sounds like a good solution, I'll go for it then.

**Comment by GuillaumeGomez** [comment]
Updated. Anything else you want me to change @ollie27 ?

File: src/librustdoc/html/layout.rs [hunk 1]
```diff
@@ -162,6 +162,8 @@ pub fn render<T: fmt::Display, S: fmt::Display>(
     css_class = page.css_class,
     logo      = if layout.logo.is_empty() {
         "".to_string()
+    } else if layout.krate.is_empty() {
+        format!("<img src='{}' alt='logo' width='100'>", layout.logo)
```
Comment (ollie27) [inline src/librustdoc/html/layout.rs]: This change is no longer needed now that the logo has been removed.

**Review by ollie27 — COMMENTED** [review]
Other than that one unneeded change this looks good to me.

**Comment by GuillaumeGomez** [comment]
Done as well!

**Pull request merged**